### PR TITLE
Fix possible memory over-read in apps/s_client.c

### PR DESCRIPTION
**Pull request by InfoHunter**
a buffer returned from BIO_gets is not checked for it's length before
reading its contents.

Signed-off-by: Paul Yang <paulyang.inf@gmail.com>

### CODE REVIEW

File: apps/s_client.c [hunk 1]
```diff
@@ -2172,6 +2172,16 @@ int s_client_main(int argc, char **argv)
              * HTTP/d.d ddd Reason text\r\n
              */
             mbuf_len = BIO_gets(fbio, mbuf, BUFSIZZ);
+            if (mbuf_len < 12) {
+                /* '12' stands for the length of 'HTTP/d.d ddd' */
```
Comment (kaduk) [inline apps/s_client.c]: Why not ```mbuf_len < strlen("HTTP/1.0 GET")```?

Comment (InfoHunter) [inline apps/s_client.c]: Errr, it would be slower to calculate the length of that string every time than just put a constant value there.

Comment (InfoHunter) [inline apps/s_client.c]: Hmmm, I get your point, the compiler may optimize the strlen("string") stuff into a constant value, isn't it?

Comment (dot-asm) [inline apps/s_client.c]: Even if compiler doesn't replace strlen("string") with constant (not all compilers do), it's not in performance-critical loop. If you are worried about it executing strlen every time you invoke s_client, then you would have to think twice, three times, four, ... hundred before invoking s_client to start with. I mean can imagine how expensive just starting program is? Or establishing TCP connection? I mean in comparison to one strlen?

Comment (InfoHunter) [inline apps/s_client.c]: Yeah, understood. Let me change this.

Comment (richsalz) [inline apps/s_client.c]: or use sizeof

Comment (InfoHunter) [inline apps/s_client.c]: Good point, let's try this.

**Comment by InfoHunter** [comment]
Updated @kaduk @dot-asm @richsalz

File: apps/s_client.c [hunk 2]
```diff
@@ -2172,6 +2172,15 @@ int s_client_main(int argc, char **argv)
              * HTTP/d.d ddd Reason text\r\n
              */
             mbuf_len = BIO_gets(fbio, mbuf, BUFSIZZ);
+            if (mbuf_len < sizeof("HTTP/1.0 200") - 1) {
```
Comment (dot-asm) [inline apps/s_client.c]: This is exactly the reason why I'd rather see strlen. It's hard to remember if sizeof("a") accounts for trailing zero or not. Code suggests that it does, but what does this print
```
int main()
{
    printf("%d\n", strlen("a"));
}
```

Comment (dot-asm) [inline apps/s_client.c]: Disregard snipped. What I'm saying is that sizeof("string") is not as intuitive as strlen.

Comment (dot-asm) [inline apps/s_client.c]: But never mind...

Comment (dot-asm) [inline apps/s_client.c]: Well, travis failed, so that needs to be addressed.

Comment (InfoHunter) [inline apps/s_client.c]: Yes, I understand what you expressed. Further more, I found a problem with using sizeof that reported by Travis-CI, the 'signed/unsigned' stuff (need explicit cast for some compilers)

For simplicity, let's use strlen instead...

Comment (dot-asm) [inline apps/s_client.c]: Just keep in mind that strlen will trigger same problem, as it returns size_t.

Comment (InfoHunter) [inline apps/s_client.c]: Okay, thanks anyway!

File: apps/s_client.c [hunk 3]
```diff
@@ -2172,6 +2172,15 @@ int s_client_main(int argc, char **argv)
              * HTTP/d.d ddd Reason text\r\n
              */
             mbuf_len = BIO_gets(fbio, mbuf, BUFSIZZ);
+            if (mbuf_len < (int)strlen("HTTP/1.0 200")) {
+                BIO_printf(bio_err,
+                           "%s: HTTP CONNECT failed, insufficient response "
+                           "from proxy (got %d octets)\n", prog, mbuf_len);
```
Comment (kaduk) [inline apps/s_client.c]: I think we tend to lean towards the "bytes" side of the bytes vs. octets debate, but given that this diagnostic should hardly ever trigger I'm not sure that it really matters.

**Comment by kaduk** [comment]
(I assume that whoever pushes will do the squashing)

**Comment by richsalz** [comment]
commit 3dce109 in master, thanks!

**Pull request closed without merging**